### ISSUE TITLE
AKS: add cluster `privatek8s`

### ISSUE THREAD

**Issue by dduportal** [jenkins-infra/helpdesk#2844]
As per https://github.com/jenkins-infra/iep/tree/master/iep-002 , we need a private AKS cluster to host the private services such as infra.ci or release.ci.

Ref infra meeting notes:
- https://github.com/jenkins-infra/documentation/blob/f8ac99c520520dead8b916aa7c6920a2e5e429ae/meetings/2021-12-21.md#notes-book
- https://github.com/jenkins-infra/documentation/blob/6f1ab2035608540b0b7ead08cb6b82314f8ade3f/meetings/2022-02-22.md

Requires https://github.com/jenkins-infra/helpdesk/issues/2683

- [x] Create a new AKS cluster `privatek8s`
  - [x] Managed in jenkins-infra/azure with Terraform
  - [x] Same settings as the current [`temp-privatek8s`](https://github.com/jenkins-infra/kubernetes-management/blob/e21c35e048c7eb41cf3959b412f8320bf1ee61d5/clusters/temp-privatek8s.yaml) which was created manually
  - [x] Should pass the Azure portal assistant analysis (security, settings, etc.) to ensure it has no "bad" settings such as wrong network CIDRs
- [x] add this cluster to kubernetes-management 
  - [x] Creates a service-account with token + insert it in infra.ci secrets
  - [x] Add initial "foundation" services (private + public ingress, cert-manager, datadog, falco)
- [x] Create a new OpenVPN (small) machine only for this service
  - [x] Infrastructure-level management in jenkins-infra/azure with Terraform
  - [x] Configuration-level management with Puppet (jenkins-infra/jenkins-infra), bootstrapped like in jenkins-infra/oracle VM
- [x] Migrate services that are obviously part of this cluster
  - [x] infra.ci from temp-privatek8s
  - [x] release.ci from prodpublick8s
    - [x] Require a Windows node pool equivalent of the current `windowspool` of the former `prodpublick8s`
    - [x] Require a"Highmem node pool equivalent of the current `highmempool` of the former `prodpublick8s`
  - [x] grafana, prometheus, loki, promtail from prodpublick8s (edit: proposal to REMOVE them until we'll need them)
  - [x] ircbot from prodpublick8s
  - [x] rss2twitter
  - [x] github-comment-ops
- [x] Decide (and migrate) if the following services have to be migrated in this cluster:
  - [x] ldap.jenkins.io => NO
  - [x] keycloak + accounts => NO
 - [x] remove `temp-privatek8s` from AKS (delete at resource group)

**Comment by dduportal** [comment]
https://github.com/jenkins-infra/helpdesk/issues/3026 => don't forget the service-principal + update the calendar notification

**Comment by timja** [comment]
we should really not be using a service principal...
it should use a managed identity instead where Microsoft will automatically rotate it

**Comment by dduportal** [comment]
@timja good points. I can't remember why did we started using an SP in the first place for this cluster. Sounds like an error we should avoid repeating there. Many thanks for the note!

**Comment by dduportal** [comment]
This issue is put in "pause":
- The cluster is no managed as code (https://github.com/jenkins-infra/azure/pull/214) and almost integrated https://github.com/jenkins-infra/kubernetes-management/pull/3247
- The infra.ci.jenkins.io migration would have been the next step
- But since https://github.com/jenkins-infra/helpdesk/issues/3227, we realized that we need to create "DualStack" networks (IPv4/IPv6) to use public IPv6.

=> Blocked by https://github.com/jenkins-infra/helpdesk/issues/2924 until we a have a new "private" network with DualStack capability.

**Comment by timja** [comment]
What's the reason for another VPN?

**Comment by dduportal** [comment]
> What's the reason for another VPN?

the 2 main reasons are:
- physical isolation : 1 vpn per network
- Cost : the current vpn VM is oversized because of the constraints of 3 NICs

The current vpn should map to the new « private VPN », and a public VPN would allow safer access on the public network for contributors who might not have to access infra.ci for instance

**Comment by timja** [comment]
What do people use the VPN for except the CI instances?

**Comment by dduportal** [comment]
> What do people use the VPN for except the CI instances?

For today:
- non admin: Accessing the CI instances (release.ci and infra.ci)
- admins: Accessing the whole private backends of the infrastructure (SSH, etc.)

We have a few new use case once this "network" is ready:

- Accessing AKS control planes through VPN only: 
  - No more publicly exposed Kubernetes API
  - Contributors of a given application (let's say plugins.jenkins.io of plugins-health.jenkins.io for instance) could be granted a safe access to their namespace to watch logs when deploying a change (or debugging)

**Comment by timja** [comment]
and why can't that be done from the existing VPN?

**Comment by dduportal** [comment]
> and why can't that be done from the existing VPN?

Connecting the new networks (with their new CIDRs) would require peering old and new networks, causing a propagation of the initial overlapping problem.

Or using another NIC on the VM but that would move the overlapping issue to the routing table of the VM.

**Comment by dduportal** [comment]
- privatek8s is up and its IP (internal LB) is reachable through the VPN private VM

**Comment by dduportal** [comment]
Next steps for 2023:
- Enable LE certificate
- Backup infra.ci.jenkins.io and migrate it
- Change DNS for infra.ci
- Add all contributors CCD in the private VPN
- Healthcheck
helathchecks

**Comment by dduportal** [comment]
We (Jenkins infra team) just deleted the following (manually managed) elements from the Azure Portal:

- AKS cluster `temp-privatek8s` (which cascade deleted the resource group `MC...temp-privatek8s...` with the LB/node pools etc.
- Service principal `temp-privatek8sSP-20211228`

**Comment by lemeurherve** [comment]
Next steps, planned for Thursday 9th of March 2023:
- [ ] **Migrate bots from `prodpublick8s` to `privatek8s`**:
	- [ ] Announce the operation on status.jenkins.io
	- [ ] Announce the operation by mail on jenkinsci-dev and jenkins-infra mailing lists
	- [ ] Announce the operation on #jenkins-infra and #jenkins-hosting IRC channels
	- [ ] Shutdown the services before their migration
	- [ ] Merge the following PRS:
		- [ ] https://github.com/jenkins-infra/kubernetes-management/pull/3527
		- [ ] https://github.com/jenkins-infra/kubernetes-management/pull/3526
		- [ ] https://github.com/jenkins-infra/kubernetes-management/pull/3525
	- [ ] Ensure the services have correctly restarted
	- [ ] Announce the end of the operation by mail on jenkinsci-dev and jenkins-infra mailing lists
	- [ ] Announce the end of the operation on #jenkins-infra and #jenkins-hosting IRC channels
	- [ ] Close the announcement on status.jenkins.io

Estimated duration: 1 to 2 hours.

- [ ] **Migrate release.ci.jenkins.io from `prodpublick8s` to `privatek8s`**:
	- [ ] Announce the operation on status.jenkins.io
	- [ ] Announce the operation by mail on jenkinsci-dev and jenkins-infra mailing lists
	- [ ] Announce the operation on #jenkins-infra and #jenkins-release IRC channels
    - [ ] Migrations steps:
    	- [ ] Merge https://github.com/jenkins-infra/azure/pull/271
    	- [ ] Add release.ci to https://github.com/jenkins-infra/kubernetes-management/blob/main/clusters/privatek8s.yaml
    		- [ ] ⚠️ Ensure no agent are available via its values so we can make potential needed changes without releasing anything by mistake
    	- [ ] Check the controller is working correctly on privatek8s
    	- [ ] Shutdown release.ci.jenkins.io
    	- [ ] Start backup of release.ci.jenkins.io
    	- [ ] Restore backup on privatek8s PV
    	- [ ] Check jobs on privatek8s
    	- [ ] Convert job-dsl to helm values like https://github.com/jenkins-infra/kubernetes-management/blob/main/config/ext_jenkins-infra-jobs.yaml (might be postponed if too short time before next Weekly release)
    	- [ ] Check jobs on privatek8s
    	- [ ] Migrate DNS
    	- [ ] Replace old IPs with the new ones in https://github.com/jenkins-infra/kubernetes-management/blob/main/config/ldap.yaml
    	- [ ] Restore agent availability for release.ci controller on privatek8s
    	- [ ] Remove release.ci from https://github.com/jenkins-infra/kubernetes-management/blob/main/clusters/prodpublick8s.yaml
	- [ ] Announce the end of the operation by mail on jenkinsci-dev and jenkins-infra mailing lists
	- [ ] Announce the end of the operation on #jenkins-infra and #jenkins-hosting IRC channels
	- [ ] Close the announcement on status.jenkins.io

Estimated duration: 1 to 2 days.

**Comment by lemeurherve** [comment]
The CNAME record propagation took longer than expected, but finally the 3 bots are back online and working as intended on their new cluster.

**Comment by lemeurherve** [comment]
Final step:
- [ ] **Migrate release.ci.jenkins.io from `prodpublick8s` to `privatek8s`**:
	- [ ] Preliminary phase:
    	- [x] Start test backup of release.ci.jenkins.io (~45min)
    	        `$ mkdir backup_releaseci_prodpublick8s && cd backup_releaseci_prodpublick8s`
				`$ kubectl config use-context prodpublick8s`
				`$ time kubectl cp release/default-release-jenkins-0:/var/jenkins_home . -c jenkins`
    	- [x] Merge https://github.com/jenkins-infra/azure/pull/271
    	- [ ] Add release.ci to https://github.com/jenkins-infra/kubernetes-management/blob/main/clusters/privatek8s.yaml
    		- [x] ⚠️ Ensure no agent are available via its values so we can make potential needed changes without releasing anything by mistake
    		- [ ] Test via a local apply: `helmfile apply -f clusters/privatek8s.yaml -l name=jenkins-release`
    	- [x] Check the controller is working correctly on privatek8s
    	- [-] ~Transform jobs as code, job-dsl to helm values like https://github.com/jenkins-infra/kubernetes-management/blob/main/config/ext_jenkins-infra-jobs.yaml (might be postponed if too short time before next Weekly release)~
    	- [ ] Restore test backup on privatek8s PV
    	        `$ cd backup_releaseci_prodpublick8s`
				`$ kubectl config use-context privatek8s`
				`$ time kubectl cp . release/default-release-jenkins-0:/var/jenkins_home -c jenkins`
    	- [ ] Apply jobs-dsl configuration and check the scan on these jobs isn't started as it should detect the existing builds from the restored backup
    	- [ ] Check jobs on privatek8s
	- [ ] Announce the operation on status.jenkins.io
	- [ ] Announce the operation by mail on jenkinsci-dev and jenkins-infra mailing lists
	- [ ] Announce the operation on #jenkins-infra and #jenkins-release IRC channels
    - [ ] Migrations steps:
    	- [ ] Shutdown release.ci.jenkins.io
    	- [ ] Start backup of release.ci.jenkins.io
    	- [ ] Restore backup on privatek8s PV
    	- [ ] Check jobs on privatek8s
    	- [ ] Migrate DNS
    	- [ ] Replace old IPs with the new ones in https://github.com/jenkins-infra/kubernetes-management/blob/main/config/ldap.yaml
    	- [ ] Restore agent availability for release.ci controller on privatek8s
    	- [ ] Remove release.ci from https://github.com/jenkins-infra/kubernetes-management/blob/main/clusters/prodpublick8s.yaml
	- [ ] Announce the end of the operation by mail on jenkinsci-dev and jenkins-infra mailing lists
	- [ ] Announce the end of the operation on #jenkins-infra and #jenkins-hosting IRC channels
	- [ ] Close the announcement on status.jenkins.io

Estimated duration: 1 to 2 days.

**Comment by lemeurherve** [comment]
The backup restoration took far more longer than expected, and I have to redo it as I want to be sure of it.
The good news is its restoration on the new cluster worked as expected, no job triggered when setting the job-dsl config part on it!

**Comment by lemeurherve** [comment]
release.ci.jenkins.io has been deployed on `privatek8s` cluster.

We're keeping the old release on `prodpublick8s` by precaution for this week releases, we're planning the cleanup for Wednesday, after the Weekly release.

**Comment by lemeurherve** [comment]
As the Weekly release succeeded and as the cleanup has been done, closing this issue.

Next steps to get rid of the manually managed `prodpublick8s` cluster described in https://github.com/jenkins-infra/helpdesk/issues/3351
